File: app/src/main/java/com/ortin/gesturetranslator/di/DataModule.java
package com.ortin.gesturetranslator.di;

import android.content.Context;

import com.ortin.gesturetranslator.data.repository.SettingsRepositoryImpl;
import com.ortin.gesturetranslator.data.storage.SettingsStorage;
import com.ortin.gesturetranslator.data.storage.SharedPrefSettingsStorage;
import com.ortin.gesturetranslator.domain.repository.SettingsRepository;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;

@Module
@InstallIn(SingletonComponent.class)
public class DataModule {
    @Provides
    @Singleton
    SettingsRepository provideSettingsRepository(SettingsStorage settingsStorage) {
        return new SettingsRepositoryImpl(settingsStorage);
    }

    @Provides
    @Singleton
    SettingsStorage provideSettingsStorage(@ApplicationContext Context context) {
        return new SharedPrefSettingsStorage(context);
    }
}

File: app/src/main/kotlin/com/ortin/gesturetranslator/app/di/DataModule.kt
package com.ortin.gesturetranslator.app.di

import com.ortin.gesturetranslator.data.repository.SettingsRepositoryImpl
import com.ortin.gesturetranslator.data.repository.UserRepositoryImpl
import com.ortin.gesturetranslator.data.storage.SettingsStorage
import com.ortin.gesturetranslator.data.storage.SharedPrefSettingsStorage
import com.ortin.gesturetranslator.domain.repository.SettingsRepository
import com.ortin.gesturetranslator.domain.repository.UserRepository
import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
interface DataModule {
    @Binds
    @Singleton
    fun bindSettingsStorage(sharedPrefSettingsStorage: SharedPrefSettingsStorage): SettingsStorage

    @Binds
    @Singleton
    fun bindSettingsRepository(impl: SettingsRepositoryImpl): SettingsRepository

    @Binds
    @Singleton
    fun bindUserRoomDatabaseRepository(impl: UserRepositoryImpl): UserRepository
}
